### PR TITLE
RFC: drivers: serial: Use microseconds to represent timeout

### LINKED ISSUE #30861: drivers: uart: increase timeout precision in uart_rx_enable
**Is your enhancement proposal related to a problem? Please describe.**
Currently, `uart_rx_enable` gets `timeout` given in milliseconds. `timeout` refers to inactivity period which leads to `UART_RX_RDY` event. Millisecond resolution may significantly lower communication throughput if protocol requires ACK. For example, device transmits 10 bytes @1Mbaud, it takes ~100us but receiver has 1ms timeout so packet will be received after ~1100us. Another case is when hardware has inactivity detection.

**Describe the solution you'd like**
Change timeout to microseconds. Optionally, add inline function or macro for converting characters to microseconds.

** Alternative **
Change timeout to characters. It is baudrate independent and may  be suitable for hardware inactivity detection.

### PR DESCRIPTION
**Pull request by nordic-krch**
Updated uart_rx_enable() and uart_tx() to use timeout given
in microseconds. Previously argument was given in milliseconds.
However, there are cases when milliseconds granularity is not
enough and can significantly reduce a throughput, e.g. 1ms is
100 bytes at 1Mb.
    
Updated 4 drivers which implement asynchronous API. Updated
places where API was used.

Additionally, added `SYS_FOREVER_US` to represent inifinity in microseconds (similar to `SYS_FOREVER_MS`).

Fixes #30861.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Comment by Mierunski** [comment]
Looks good, just wondering if we should switch to `int64_t` in tx/rx_timeout

**Review by Mierunski — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]
LGTM. Since this is changing a stable API, this needs to go through the stable API change process.

**Comment by nordic-krch** [comment]
@dcpleung According to this table https://docs.zephyrproject.org/latest/reference/api/overview.html it is still unstable.
@carlescufi Can you add this to API meeting agenda?

**Comment by nordic-krch** [comment]
> wondering if we should switch to int64_t in tx/rx_timeout

It's >4200 seconds. To me it sounds enough but can be changed to 64 bits if there are any concerns.

**Comment by dcpleung** [comment]
Ah... the UART async API is still unstable, but the UART API is. So not an issue then.

**Comment by cfriedt** [comment]
@nordic-krch Can this be merged or is it still an RFC?

**Comment by nordic-krch** [comment]
@cfriedt let's discuss it on api meeting today.

File: include/drivers/uart.h [hunk 1]
```diff
@@ -486,8 +486,8 @@ static inline int uart_callback_set(const struct device *dev,
  * @param dev     UART device structure.
  * @param buf     Pointer to transmit buffer.
  * @param len     Length of transmit buffer.
- * @param timeout Timeout in milliseconds. Valid only if flow control is
```
Comment (carlescufi) [inline include/drivers/uart.h:489]: Open question: should we use `int64_t` to allow for longer timeouts?

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -808,10 +808,10 @@ static inline void async_evt_rx_buf_release(struct uart_stm32_data *data)
 static inline void async_timer_start(struct k_work_delayable *work,
 				     int32_t timeout)
 {
-	if ((timeout != SYS_FOREVER_MS) && (timeout != 0)) {
+	if ((timeout != SYS_FOREVER_US) && (timeout != 0)) {
 		/* start timer */
 		LOG_DBG("async timer started for %d ms", timeout);
```
Comment (anangl) [inline drivers/serial/uart_stm32.c:813]:
```suggestion
		LOG_DBG("async timer started for %d us", timeout);
```

File: drivers/serial/uart_nrfx_uarte.c [hunk 3]
```diff
@@ -1044,11 +1044,11 @@ static void rxstarted_isr(const struct device *dev)
...
 	};
 	user_callback(dev, &evt);
-	if (data->async->rx_timeout != SYS_FOREVER_MS) {
+	if (data->async->rx_timeout != SYS_FOREVER_US) {
 		data->async->rx_timeout_left = data->async->rx_timeout;
 		k_timer_start(&data->async->rx_timeout_timer,
-			      K_MSEC(data->async->rx_timeout_slab),
-			      K_MSEC(data->async->rx_timeout_slab));
+			      K_USEC(data->async->rx_timeout_slab),
+			      K_USEC(data->async->rx_timeout_slab));
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:1051]: Nit: two comments, in lines [929](https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L929) and [999](https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L999), still mention "ms" as the RX timeout unit.

Also the calculation of `rx_timeout_slab`:
https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L809-L811
 should be updated, as the minimum value there is set for milliseconds.

File: include/drivers/uart.h (hunk 1 above)
Comment (anangl) [reply include/drivers/uart.h:489]: I think there is no need to. With `int32_t`, the maximum value gives almost 36 minutes, so it seems to be enough for UART communication.

**Review by carlescufi — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**